File: frontend/src/utils/ical.js
/**
 * 课表 iCal (.ics) 导出工具
 * 生成符合 RFC 5545 的日历文件
 */

import { getPeriodTime } from './periods'

// RFC 5545 日期时间格式
const fmt = {
  DTSTART: (d) => formatICSDate(d),
  DTEND:   (d) => formatICSDate(d),
  DTSTAMP: ()  => formatICSDate(new Date()),
}

/** 将 Date 转为 ICS 格式 YYYYMMDDTHHMMSS */
function formatICSDate(date) {
  const pad = n => String(n).padStart(2, '0')
  const y = date.getFullYear()
  const mo = pad(date.getMonth() + 1)
  const d = pad(date.getDate())
  const h = pad(date.getHours())
  const mi = pad(date.getMinutes())
  const s = pad(date.getSeconds())
  return `${y}${mo}${d}T${h}${mi}${s}`
}

/** 转义 ICS 特殊字符 */
function escapeICS(str) {
  if (!str) return ''
  return String(str)
    .replace(/\\/g, '\\\\')
    .replace(/;/g, '\\;')
    .replace(/,/g, '\\,')
    .replace(/\n/g, '\\n')
}

/** 获取课程下一节课的日期 */
function getNextCourseDate(dayOfWeek, periodStart) {
  const now = new Date()
  const dayMap = [0, 1, 2, 3, 4, 5, 6] // JS: 0=周日
  const targetJS = dayOfWeek === 7 ? 0 : dayOfWeek

  const next = new Date(now)
  next.setHours(0, 0, 0, 0)
  const diff = (targetJS - now.getDay() + 7) % 7 || 7
  next.setDate(now.getDate() + diff)

  const timeMap = {
    1:  [8,  0], 2:  [8,  55], 3:  [10, 0], 4:  [10, 55],
    5:  [14, 0], 6:  [14, 55], 7:  [15, 50], 8:  [16, 45],
    9:  [18, 30], 10: [19, 25], 11: [20, 20], 12: [21, 15],
  }
  const [h, m] = timeMap[periodStart] || [8, 0]
  next.setHours(h, m, 0, 0)
  return next
}

/**
 * 生成 iCal 文件内容
 * @param {Object} scheduleData - 课表数据
 * @param {string} studentName - 学生姓名
 * @returns {string} .ics 文件文本
 */
export function generateICal(scheduleData, studentName = '') {
  const uidBase = Date.now()
  const lines = [
    'BEGIN:VCALENDAR',
    'VERSION:2.0',
    'PRODID:-//JW System//课表导出//ZH',
    'CALSCALE:GREGORIAN',
    'METHOD:PUBLISH',
    'X-WR-CALNAME:' + escapeICS((studentName ? studentName + ' ' : '') + '课程表'),
    'X-WR-TIMEZONE:Asia/Shanghai',
    'X-APPLE-CALENDAR-COLOR:#007AFF',
  ]

  const courses = scheduleData?.courses || []
  const seenKeys = new Set()
  let uidCounter = 0

  courses.forEach(course => {
    // 按 (dayOfWeek, periodStart, name) 去重
    const key = `${course.dayOfWeek}-${course.periodStart}-${course.name}`
    if (seenKeys.has(key)) return
    seenKeys.add(key)

    const startDate = getNextCourseDate(course.dayOfWeek, course.periodStart)
    const periodTime = getPeriodTime(course.periodStart)
    const endMinutes = course.periods * 55
    const endDate = new Date(startDate.getTime() + endMinutes * 60 * 1000)

    // BYDAY: MO/TU/WE/TH/FR/SA/SU
    const dayNames = ['SU', 'MO', 'TU', 'WE', 'TH', 'FR', 'SA']
    const byDay = dayNames[course.dayOfWeek === 7 ? 0 : course.dayOfWeek]

    // RRULE: 直到 2025-08-01 学期结束
    const until = '20250801T000000Z'

    const uid = `${uidBase}-${uidCounter++}@jw-system`

    lines.push('BEGIN:VEVENT')
    lines.push(`UID:${uid}`)
    lines.push(`DTSTAMP:${fmt.DTSTAMP()}`)
    lines.push(`DTSTART:${fmt.DTSTART(startDate)}`)
    lines.push(`DTEND:${fmt.DTEND(endDate)}`)
    lines.push(`RRULE:FREQ=WEEKLY;BYDAY=${byDay};UNTIL=${until}`)
    lines.push(`SUMMARY:${escapeICS(course.name)}`)
    lines.push(`LOCATION:${escapeICS(course.room || '')}`)
    lines.push(`DESCRIPTION:${escapeICS(course.teacher ? '教师: ' + course.teacher : '')}`)
    lines.push(`CATEGORIES:${escapeICS(course.room ? '课程' : '')}`)
    lines.push('STATUS:CONFIRMED')
    lines.push('TRANSP:OPAQUE')
    lines.push('END:VEVENT')
  })

  lines.push('END:VCALENDAR')
  return lines.join('\r\n')
}

/**
 * 触发浏览器下载 .ics 文件
 */
export function downloadICal(content, filename = 'schedule.ics') {
  const blob = new Blob([content], { type: 'text/calendar;charset=utf-8' })
  const url = URL.createObjectURL(blob)
  const link = document.createElement('a')
  link.href = url
  link.download = filename
  document.body.appendChild(link)
  link.click()
  document.body.removeChild(link)
  URL.revokeObjectURL(url)
}

File: internal/handler/ical.go
package handler

import (
	"crypto/rand"
	"encoding/base64"
	"fmt"
	"net/http"
	"strings"
	"sync"
	"time"

	"github.com/gin-gonic/gin"
	"jww/internal/model"
	"jww/internal/service"
	"jww/pkg/response"
)

// iCalTokenStore 内存存储 iCal token（token -> sessionID）
var (
	iCalTokenStore = make(map[string]*iCalTokenEntry)
	iCalMu         sync.RWMutex
)

type iCalTokenEntry struct {
	SessionID  string
	ExpireTime time.Time
}

// GenerateICalToken 生成 90 天订阅 token
func (h *ScheduleHandler) GenerateICalToken(c *gin.Context) {
	sessionIDStr, ok := getSessionID(c)
	if !ok {
		response.Error(c, http.StatusUnauthorized, "无效的会话")
		return
	}

	// 生成随机 token
	tokenBytes := make([]byte, 32)
	rand.Read(tokenBytes)
	token := base64.URLEncoding.EncodeToString(tokenBytes)

	// 存储，90 天过期
	iCalMu.Lock()
	iCalTokenStore[token] = &iCalTokenEntry{
		SessionID:  sessionIDStr,
		ExpireTime: time.Now().Add(90 * 24 * time.Hour),
	}
	iCalMu.Unlock()

	// 返回订阅 URL
	subscribeURL := fmt.Sprintf("%s/api/schedule/ical/subscribe?token=%s", getBaseURL(c), token)
	response.Success(c, gin.H{
		"token":    token,
		"url":      subscribeURL,
		"webcal":   "webcal://" + strings.TrimPrefix(subscribeURL, "https://"),
		"expireAt": time.Now().Add(90 * 24 * time.Hour).Format("2006-01-02"),
	})
}

// SubscribeICal 通过 token 免登录订阅 iCal
func (h *ScheduleHandler) SubscribeICal(c *gin.Context) {
	token := c.Query("token")
	if token == "" {
		c.JSON(400, gin.H{"error": "缺少 token 参数"})
		return
	}

	iCalMu.RLock()
	entry, exists := iCalTokenStore[token]
	iCalMu.RUnlock()

	if !exists || time.Now().After(entry.ExpireTime) {
		c.JSON(401, gin.H{"error": "token 已过期或无效"})
		return
	}

	// 获取课表
	jwSvc := service.GetJwService()
	fullSchedule, err := jwSvc.GetFullSchedule(entry.SessionID, 20)
	if err != nil || fullSchedule == nil {
		c.JSON(500, gin.H{"error": "获取课表失败"})
		return
	}

	// 生成 iCal
	ical := generateICalContent(fullSchedule)

	c.Header("Content-Type", "text/calendar; charset=utf-8")
	c.Header("Content-Disposition", fmt.Sprintf("attachment; filename=\"课程表-%s.ics\"", fullSchedule.StudentName))
	c.String(200, ical)
}

// GetICal 获取当前用户的 iCal（需登录）
func (h *ScheduleHandler) GetICal(c *gin.Context) {
	sessionIDStr, ok := getSessionID(c)
	if !ok {
		response.Error(c, http.StatusUnauthorized, "无效的会话")
		return
	}

	jwSvc := service.GetJwService()
	fullSchedule, err := jwSvc.GetFullSchedule(sessionIDStr, 20)
	if err != nil || fullSchedule == nil {
		response.Error(c, http.StatusInternalServerError, "获取课表失败")
		return
	}

	ical := generateICalContent(fullSchedule)

	c.Header("Content-Type", "text/calendar; charset=utf-8")
	c.Header("Content-Disposition", fmt.Sprintf("attachment; filename=\"课程表-%s.ics\"", fullSchedule.StudentName))
	c.String(200, ical)
}

// GetICalTokenInfo 获取当前用户的 token 信息
func (h *ScheduleHandler) GetICalTokenInfo(c *gin.Context) {
	sessionIDStr, ok := getSessionID(c)
	if !ok {
		response.Error(c, http.StatusUnauthorized, "无效的会话")
		return
	}

	iCalMu.RLock()
	defer iCalMu.RUnlock()

	var tokenInfo *struct {
		Token    string `json:"token"`
		ExpireAt string `json:"expireAt"`
	}

	for token, entry := range iCalTokenStore {
		if entry.SessionID == sessionIDStr && time.Now().Before(entry.ExpireTime) {
			tokenInfo = &struct {
				Token    string `json:"token"`
				ExpireAt string `json:"expireAt"`
			}{
				Token:    token,
				ExpireAt: entry.ExpireTime.Format("2006-01-02"),
			}
			break
		}
	}

	if tokenInfo == nil {
		response.Success(c, nil)
		return
	}
	response.Success(c, tokenInfo)
}

// generateICalContent 生成 iCalendar 格式内容
func generateICalContent(schedule *model.FullSchedule) string {
	var sb strings.Builder
	sb.WriteString("BEGIN:VCALENDAR\r\n")
	sb.WriteString("VERSION:2.0\r\n")
	sb.WriteString("PRODID:-//jww.p//Course Schedule//CN\r\n")
	sb.WriteString("CALSCALE:GREGORIAN\r\n")
	sb.WriteString("METHOD:PUBLISH\r\n")
	sb.WriteString("X-WR-CALNAME:课程表\r\n")
	sb.WriteString("X-WR-TIMEZONE:Asia/Shanghai\r\n")

	// 计算学期第一天（如果已知）
	semesterStart := schedule.SemesterStart
	if semesterStart == "" {
		semesterStart = time.Now().Format("2006-01-02")
	}
	startDate, _ := time.Parse("2006-01-02", semesterStart)

	for _, course := range schedule.Courses {
		for _, week := range course.Weeks {
			// 计算这门课在这一周的日期
			courseDate := startDate.AddDate(0, 0, (week-1)*7+(course.DayOfWeek-1))
			startTime := getPeriodStartTime(course.PeriodStart)
			endTime := getPeriodEndTime(course.PeriodStart, course.Periods)

			dtStart := fmt.Sprintf("%sT%s", courseDate.Format("20060102"), startTime)
			dtEnd := fmt.Sprintf("%sT%s", courseDate.Format("20060102"), endTime)

			sb.WriteString("BEGIN:VEVENT\r\n")
			sb.WriteString(fmt.Sprintf("UID:%s-%d-%d@jwschedule\r\n", course.Name, week, course.DayOfWeek))
			sb.WriteString(fmt.Sprintf("DTSTART:%s\r\n", dtStart))
			sb.WriteString(fmt.Sprintf("DTEND:%s\r\n", dtEnd))
			sb.WriteString(fmt.Sprintf("SUMMARY:%s\r\n", escapeICalText(course.Name)))
			sb.WriteString(fmt.Sprintf("LOCATION:%s\r\n", escapeICalText(course.Room)))
			sb.WriteString(fmt.Sprintf("DESCRIPTION:%s\\n教师: %s\\n第%d周/周%d/第%d节\r\n",
				escapeICalText(course.Name), escapeICalText(course.Teacher), week, course.DayOfWeek, course.PeriodStart))
			sb.WriteString(fmt.Sprintf("RRULE:FREQ=WEEKLY;COUNT=%d;BYDAY=%s\r\n",
				len(course.Weeks), getWeekdayICal(course.DayOfWeek)))
			sb.WriteString("END:VEVENT\r\n")
		}
	}

	sb.WriteString("END:VCALENDAR\r\n")
	return sb.String()
}

func getWeekdayICal(dayOfWeek int) string {
	days := []string{"MO", "TU", "WE", "TH", "FR", "SA", "SU"}
	if dayOfWeek >= 1 && dayOfWeek <= 7 {
		return days[dayOfWeek-1]
	}
	return "MO"
}

func getPeriodStartTime(periodStart int) string {
	// 前两节 8:00，后续每节 45 分钟
	times := []string{"080000", "081500", "090000", "091500", "100000", "101500", "110000", "111500",
		"140000", "141500", "150000", "151500", "160000", "161500", "170000", "171500",
		"190000", "191500", "200000", "201500", "210000", "211500", "220000", "221500"}
	idx := periodStart - 1
	if idx < 0 {
		idx = 0
	}
	if idx >= len(times) {
		idx = len(times) - 1
	}
	return times[idx]
}

func getPeriodEndTime(periodStart, periods int) string {
	// 简单计算：每节课 45 分钟
	startMinutes := (periodStart - 1) * 45
	endMinutes := startMinutes + periods*45
	hour := 8 + endMinutes/60
	minute := endMinutes % 60
	return fmt.Sprintf("%02d%02d00", hour, minute)
}

func escapeICalText(s string) string {
	s = strings.ReplaceAll(s, "\\", "\\\\")
	s = strings.ReplaceAll(s, ";", "\\;")
	s = strings.ReplaceAll(s, ",", "\\,")
	s = strings.ReplaceAll(s, "\n", "\\n")
	return s
}

func getBaseURL(c *gin.Context) string {
	scheme := "https"
	if c.Request.TLS == nil {
		scheme = "http"
	}
	return scheme + "://" + c.Request.Host
}
